src/System.Diagnostics.Process/tests/ProcessStreamReadTests.cs

```diff
@@ -88,6 +88,243 @@ public void TestAsyncOutputStream()
+            Console.WriteLine(++counter);
+            Console.WriteLine(++counter);
+            Console.WriteLine(++counter);
+
+            // Signal 4,5,6 produced
+            Assert.True(childLockList["childProduced456"].TryAcquire(WaitInMS), "Child process should acquire 'childProduced456' lock");
+
+            // Wait second BeginOutputReadLine() from parent
+            Assert.True(childLockList["parentSecondBeginOutputReadLine"].WaitSignal(WaitInMS), "Parent signal missing");
+
+            // Produce 7,8,9
+            Console.WriteLine(++counter);
+            Console.WriteLine(++counter);
+            Console.WriteLine(++counter);
+
+            // Signal 7,8,9 produced
+            Assert.True(childLockList["childProduced789"].TryAcquire(WaitInMS), "Child process should acquire 'childProduced789' lock");
+
+            // Shutdown process
+            Assert.True(childLockList["shutdownChildProcess"].WaitSignal(WaitInMS), "Signal to shutdown not come after 'WaitInMS'");
+
+            return SuccessExitCode;
+        }
+
+        private Dictionary<string, FileSystemLock> TestAsyncOutputStream_BeginCancelBegin_OutputReadLine_LockHelper(string directoryName)
+        {
+            return new Dictionary<string, FileSystemLock>
+            {
+                { "childStarted", new FileSystemLock("childStarted", directoryName) },
+                { "parentFirstBeginOutputReadLine", new FileSystemLock("parentFirstBeginOutputReadLine", directoryName) },
+                { "parentSecondBeginOutputReadLine", new FileSystemLock("parentSecondBeginOutputReadLine", directoryName) },
+                { "parentCancelOutputRead", new FileSystemLock("parentCancelOutputRead", directoryName)},
+                { "childProduced123", new FileSystemLock("childProduced123", directoryName) },
+                { "childProduced456", new FileSystemLock("childProduced456", directoryName) },
+                { "childProduced789", new FileSystemLock("childProduced789", directoryName) },
+                { "shutdownChildProcess", new FileSystemLock("shutdownChildProcess", directoryName) }
+            };
+        }
+
+        /// <summary>
+        /// We save lock files in temp directory
+        /// </summary>
+        private class FileSystemLock : IDisposable
+        {
+            private readonly string _lockFileName = "";
+            private FileStream _fs = null;
+
+            public FileSystemLock(string lockName, string directoryName)
+            {
+                string rootDirectory = Path.Combine(Path.GetTempPath(), directoryName);
+
+                if (!Directory.Exists(rootDirectory))
+                {
+                    Directory.CreateDirectory(rootDirectory);
+                }
+
+                _lockFileName = Path.Combine(rootDirectory, lockName);
+            }
+
+            public bool WaitSignal(int millisecondsTimeout)
+            {
+                DateTime maxWait = DateTime.UtcNow.AddMilliseconds(millisecondsTimeout);
```

Comment:
Nit, no need to change: when timing millisecond level, it would be more idiomatic (and simpler) to use Stopwatch.

Reply:
fixed thank's!